### PR TITLE
Fix #47753 - Organize imports removes type imports that are only referenced in @link (jsdoc)

### LINKED ISSUE #47753: Organize imports removes type imports that are only referenced in `@link` (jsdoc)
# Bug Report

### 🔎 Search Terms

organize imports link

### 🕗 Version & Regression Information

As far as I know this is the behavior ever since `@link` support was added.

### ⏯ Playground Link

[Playground link with relevant code][pg]

### 💻 Code

```ts
/// <refenrece types="zod"/>
import type { ZodType } from 'zod';

/** Intended to be used in combination with {@link ZodType} */
export function fun() { /* ... */ }
```

### 🙁 Actual behavior

Import is shown as not being used and gets removed on executing "Organize imports".

### 🙂 Expected behavior

Keep it.

[pg]: https://www.typescriptlang.org/play?#code/PQgEB4CcFMDNoHYwMbVAFwJ4AdoGcBeAIgC8B7AEyOAD4BYAKAEsBbbMydDHNAb1ABalACo9QAX1CxIZFqADk5CvIDcjRsABUm0AEkE6RBWgUMZUACM0AVzwnQTBKGSyLjgIbomZJwHcm6AAWoLwAAgA2jgDWgiI8kprAjNAAHuycUtYIyF4+mQgAFACUIeoM4oxAA

### PR DESCRIPTION
**Pull request by komyg**
Fixes #47753, https://github.com/microsoft/TypeScript/issues/47558

### CODE REVIEW

File: src/services/organizeImports.ts [hunk 1]
```diff
@@ -164,7 +165,29 @@ namespace ts.OrganizeImports {
         function isDeclarationUsed(identifier: Identifier) {
             // The JSX factory symbol is always used if JSX elements are present - even if they are not allowed.
             return jsxElementsPresent && (identifier.text === jsxNamespace || jsxFragmentFactory && identifier.text === jsxFragmentFactory) && jsxModeNeedsExplicitImport(compilerOptions.jsx) ||
-                FindAllReferences.Core.isSymbolReferencedInFile(identifier, typeChecker, sourceFile);
+                FindAllReferences.Core.isSymbolReferencedInFile(identifier, typeChecker, sourceFile) ||
```
Comment (sandersn) [inline src/services/organizeImports.ts:168]: organiseImports should rely on FindAllReferences to determine whether a link tag is referenced.
I'm not 100% sure how it should work, but I think that FindAllReferences, when determining whether `ZodType` is referenced, should

1. find the text `ZodType` in the text of an `@link` tag. (This should already be the case).
2. Use checker.getSymbolAtLocation to resolve the `link.name` to a symbol. I don't know what FindAllReferences normally uses to resolve names. You might have to add a special case here for when the identifier is a `link.name`.
3. If the name resolves, then it is referenced and `isSymbolReferencedInFile` should return true.

**Review by sandersn — CHANGES_REQUESTED** [review]
This works, but it should actually happen in FindAllReferences, and use the normal FindAllReference machinery.

**Comment by komyg** [comment]
Hi @sandersn, thanks for the review. I will update the PR accordingly, but I just need a little time, because things are kind of crazy right now.

File: src/services/organizeImports.ts (hunk 1 above)
Comment (komyg) [reply src/services/organizeImports.ts:168]: I have a question about this: I saw that the find all references has two methods that work with JSDocs: `getParameterSymbolFromJSDoc` and `getTypeParameterFromJsDoc`. From what I could gather, both methods traverse the tree to find the appropriate symbol for a given JSDoc element.

For example, the `getParameterSymbolFromJSDoc` is used for `@param` tags and checks the function declaration that contains this JSDoc to get the Symbol from the function's arguments. 

Something similar happens with the `getTypeParameterFromJsDoc`  function (I think).

My point is that we are not creating new Symbols, but retrieving them from the existing elements of the tree and returning them.

However, in the case of my PR, we expect that the Symbol for the `{@link}` tag is on the top of the tree on the import declarations. So do I have to traverse all the way up the tree to get the correct Symbol, or do you think there is a better way to do this?

**Comment by DanielSWolf** [comment]
This is related to the larger issue that TS considers `@link`ed imports as unused, as described in #47558. I wonder whether it might make more sense to solve the root cause rather than fixing this specific symptom.

**Comment by komyg** [comment]
Hey @DanielSWolf, I think I found the root cause of both problems in the compiler. I will try to solve both issues.

**Comment by DanielSWolf** [comment]
@komyg That would be great! The current behavior makes it really hard to write links that consistently work.

**Comment by komyg** [comment]
Hi @sandersn, I've fixed the `FindAllReference` function as you requested and all test cases for organize imports have passed.

However, I also got an error for a JSDoc baseline that I didn't understand, so I don't know if I inadvertently broke something or if the fix also corrected it.

Could you please review the PR again?

**Comment by sandersn** [comment]
WIthout debugging the newly-failing test case, I don't know how the old code resolves `C` to the original class. What happens if you change `/*dontResolveAlias*/ !isJSDoc` to `/*dontResolveAlias*/ false`? It's possible that it doesn't actually *ever* need to be true.

**Comment by komyg** [comment]
> WIthout debugging the newly-failing test case, I don't know how the old code resolves `C` to the original class. What tests fail if you change `/*dontResolveAlias*/ !isJSDoc` to `/*dontResolveAlias*/ false`? It's possible that it doesn't actually _ever_ need to be true.

Well, I since the `dontResolveAlias` param is `!JSDoc` it will probably be always `false`. 

I think that the problem lies in the `resolveEntityName` function. I've followed its execution through the debugger, and it resolves the symbol for the `{@link ZodType}`, however at its end there is the following conditional:

```ts
return (symbol.flags & meaning) || dontResolveAlias ? symbol : resolveAlias(symbol);
```

By setting `dontResolveAlias` to `true` or by adding the `Symbol.Alias` meaning when calling `resolveEntityName`, we fall into the first part of the ternary and just return the symbol we found. Which in our case means the `import` statement that is referenced by the JSDoc `{@link ZodType}` and the import from the `C` type in the failing test.

Since the original code did not have the `Symbol.Alias` in the `meaning` and `dontResolveAlias` was `false`, we fell into the second part of the ternary, which meant we tried to resolve our JSDoc `{@link ZodType}` import alias twice, resulting in an unknown type and the bug that was reported.

A possible solution would be for us to find what is unique about each situation to make a better choice whether to set `dontResolveAlias` to `true` or `false` or to update the `meaning` accordingly. But I don't know if this is possible, since we have two different use cases of the same `{@link}` element.

**Comment by komyg** [comment]
How can we fix this, then? It seems that _organize imports_ is an "edgier" case than _resolving aliases_ that you mentioned, so maybe we could do some logic outside the `findAllReferences` code.

My original solution did not rely on the `findAllReferences` code. Maybe we can repurpose that somehow or add it to the checker code? What do you think?

Note: this is the [commit](https://github.com/microsoft/TypeScript/pull/47824/commits/229d85dc16b95cd95c1331b0fd3cd1210e83b943) in which I erased my previous attempt at a solution.

**Comment by sandersn** [comment]
I looked again at your test cases, and the first problem is that they import things that aren't present in the program. The test cases need to have multiple files. See jsdocLink3.ts for an example.

It might be nice if organise imports didn't rely on find-all-refs, and therefore on the the target of the alias being defined.

**Comment by sandersn** [comment]
From manual testing, I can see that find-all-references doesn't work correctly for `@link` references, so it should be fixed too. I still believe that fixing find-all-refs will fix organise-imports. Here's the test I used:

```ts
// @filename:ex.d.ts
export type A = 1 | 2
export type B = 3 | 4
// @filename: main.ts
import { A, B } from './ex'
/** {@link A} */
class C {
  b: B = 3
}
```

Find-all-refs on main's B finds 3 refs; find-all-refs on main's A only finds the import alias.

**Comment by sandersn** [comment]
> WIthout debugging the newly-failing test case, I don't know how the old code resolves C to the original class. What tests fail if you change /*dontResolveAlias*/ !isJSDoc to /*dontResolveAlias*/ false? It's possible that it doesn't actually ever need to be true.

I tested this locally and it is the solution. It does change the baselines for jsdocLink3, but I'm pretty sure those were wrong, because I manually tested the test code in jsdocLink3 and it behaves right with `false`, wrong with `!isJSDoc`. When I wrote the code, I didn't understand how it was supposed to follow (or not) aliases.

**Comment by komyg** [comment]
Hey @sandersn, from your comments, I understood that the solution that we've implemented here is correct, and I just have to make the following changes:

1. Update my test cases, so that they have types defined in multiple files, as per your example and the _jsdocLink3.ts_ test.

2. Change the call to `findAllReferences` so that the `dontResolveAlias` is always `false`, but also keep the change that I made that adds `Symbol.Alias` to the `meaning`.

3. Update the _jsdocLink3.ts_ baseline.

**Comment by sandersn** [comment]
Sorry, **true**. dontResolveAlias should be **true**.

**Comment by komyg** [comment]
Hi @sandersn, I've done the changes. Could you review again, please?

**Comment by komyg** [comment]
Hi @sandersn, sorry to bother you. I've done the changes we've discussed. Could you review again, please?

**Review by sandersn — APPROVED** [review]

**Comment by sandersn** [comment]
Woops, sorry this fell off my list. Thanks for your patience!

**Comment by komyg** [comment]
No worries! I am happy to have contributed!

**Pull request merged**